### ISSUE TITLE
Docker container links are obsoleted - must use networks.

### ISSUE THREAD

**Issue by flukejones** [rockstor/rockstor-core#1982]
I recently tried translating a fairly standard owncloud docker-compose to Rockon format.

* Container links have no effect,
* Container links are [being phased out ](https://docs.docker.com/network/links/)

For containers to communicate without exporting ports they must be connected to the same user-created network. That is, non-default network. This is especially critical if you want to be able to use DNS resolution to make container communications easier.

The best thing to do would be to retain the existing container linking json, and use that to determine how to network:

```
    "container_links": {
      "container-A": [
        {
          "name": "link-B",
          "source_container": "container-B"
        },
        {
          "name": "link-C",
          "source_container": "container-C"
        }
      ]
    },
```
Using the above *A* needs to see *B* and *C*, but *B* and *C* don't need to see each other. The docker commands would simply be:

```
docker network create link-B
docker network connect container-A link-B
docker network connect container-B link-B
// create network if not exists
docker network create link-C
docker network connect container-A link-C
docker network connect container-C link-C
```
And for:

```
    "container_links": {
      "container-A": [
        {
          "name": "link-B",
          "source_container": "container-B"
        },
        {
          "name": "link-B",
          "source_container": "container-C"
        }
      ]
    },
```

Which means all three containers are in the same net. Since it just means: 

```
docker network create link-B
docker network connect container-A link-B
docker network connect container-B link-B
// create network if not exists
docker network connect container-C link-B
```

**Comment by phillxnet** [comment]
@Luke-Nukem  Thanks again for another nice exposition and for the heads up.

As this looks to depends on Rockstor's Rock-on system being extended to be aware of the 'docker network' element I'll link to a prior forum thread where this was previously raised:

https://forum.rockstor.com/t/rockon-networking/3077/8

There I sketched out the UI elements that I thought at the time might be required to support the docker network capability. Note however that I'm not that up on docker myself so all informed input is welcome. Maybe you could contribute further to that thread so that we might work out how this would best be implemented from the UI perspective whilst still maintaining ease of use / simplicity of deployment. 

And by way of a quote from your referenced link:
"One feature that user-defined networks do not support that you can do with --link is sharing environmental variables between containers."
I'll link to the issues opened by @daniel-illi :
"Rockons: map configured env variable to multiple containers" #1589
and:
"Rockons: config options for hidden env variables and volume mounts" #1591

As always we have to build these capabilities bit by bit whilst trying not to break existing functionality / rock-on definitions.

Please consider looking to the above links and also taking a look at the Rock-on code within Rockstor and extending it's docker capabilities as you suggest. It is definitely due for some more recent attention and in fact has had a couple of updates of late by @FroggyFlox in the following 2 pull requests that may also serve to point you at the appropriate code, at least for the Rock-on Web-UI elements:
"Add support for mounting devices to Rockons. Fixes #1957"
and
"Add Support for Docker run command arguments. Fixes #1967"

Thanks again and I see that you have already joined the forum so hope to engage further with you there also on developing Rockstor's current capabilities. But note that we do have several pending issues that relate to basic functionality that are keeping at least me fairly busy (at least for the time being) so do consider chipping in where you can on anything that takes your fancy. Ultimately all pull requests are reviewed in turn by project lead @schakrava, although @FroggyFlox is the most recent contributor to the Rock-on code itself.

**Comment by FroggyFlox** [comment]
Hi @Luke-Nukem and @phillxnet ,

I'll share my thoughts on this as I've been thinking about the whole Rock-on system quite a bit lately. I'm not familiar enough with "advanced" Docker features and not up-to-date either, so thank you very much @Luke-Nukem for bringing this out and your recommendations!

Regarding the `container_links` object we have in the Rock-on definition, I recently realized it was simply ignored by the general Rock-on install process and seemed to be the remnant of older custom functions. Here's the forum thread that made me realize it: https://forum.rockstor.com/t/nextcloud-rockon-with-postgres/5255/6?u=flox. Based on this, my initial thought was to deprecate the `container_links` object altogether for simplicity (as the `opts` object can be used instead). I do like @Luke-Nukem 's suggestion to use it to create docker networks accordingly, however, and it would seem to be an elegant way to deal with the `container_links`.

@phillxnet , thanks for pointing the forum thread with your suggestions re Docker networking, it seems highly relevant to the problem and I'll most likely share my thoughts about its implementation there as well. Briefly--and in the way I see the Rock-On system's "philosophy"--I usually tend to believe that Rockstor would not necessarily benefit much from an implementation of advanced Docker features as these would most likely be quite complex to do correctly (from a user-friendly and simplicity perspective). Users who would know how to configure such features would most likely be better running the container(s) in question manually (or with `docker compose` for instance). For me, Rock-Ons were my first encounter with the Docker ecosystem and it is the extreme simplicity of the current Rock-Ons that led me to realize the simplicity and benefit to using Docker itself, even though I never ran Docker containers before. If their setup were complex, I probably would not have tried it so willingly.

With this in mind, I was thinking of a way to keep the Rock-Ons installs as straightforward as possible:  
1. Implement @Luke-Nukem's suggestion to use the `container_links` object to define docker networks. This would be mostly useful for multi-containers Rock-Ons (on this topic, thanks @Luke-Nukem for your commenting on #1588)
2. Post-installation, offer the possibility to create a new docker network in the "System > Network" sub-menu, and join an existing docker network from the "customize rock-on" panel.

This should cover most use-cases, isn't it?

Let me know if this is an overly complicated way to look at the Rock-On system, I'm curious to know whether that is the actual philosophy behind it or whether I simply made this up in my mind, :-)

PS: Regarding point 2 above, I'm currently trying to work on a way to offer just this: adding your own `label`, `device`, etc... based on the current "Add Storage" feature. I'm moving extremely slowly on it, but that could be used to customize docker networks as well...

**Comment by Luke-Nukem** [comment]
I figure it best to reply here to keep the new discussion centralised.

>1) Implement [@Luke-Nukem](https://github.com/Luke-Nukem)'s suggestion to use the `container_links` object to define docker networks. This would be mostly useful for multi-containers Rock-Ons (on this topic, thanks [@Luke-Nukem](https://github.com/Luke-Nukem) for your commenting on [#1588](https://github.com/rockstor/rockstor-core/issues/1588))
>2) Post-installation, offer the possibility to create a new docker network in the "System &gt; Network" sub-menu, and join an existing docker network from the "customize rock-on" panel.

Generally speaking, creating extra networks for containers (or composes) that have already created them isn't required as the docker networking and DNS is robust. The only reason to create new networks is to connect containers that haven't exported their ports, only exposed them (meaning they are hidden from everything not in the same network). A good example of this is the owncloud/server compose. In fact the owncloud compose demonstrates the need to have the networking solution I outline, and thus prevent the need for people to create AIO containers that eventually become out of date and unmaintained.

The example of using nginx as a reverse proxy is also only required if containers haven't exported their ports - otherwise nginx can just use the local host IP:PORT combo. Speaking of which, I would actually like to see an option to export or not the ports in the GUI, *then* we would have a use for network creation and joining. Also using the nginx reverse proxy example - you would be joining nginx to the networks conatiner JSON/composes have created, not created new networks.

Lastly... it's good practice for newly created containers or composes to be on their own networks anyway, not on the docker defaults as the defaults don't have DNS (meaning you can link containers using names, not IP, including in nginx configs).

I'm going to say that this whole issue is a critical one, if not just for the fact that people are creating all-in-one containers to solve some of these issues.

**Comment by baforgood** [comment]
Dear all,

do I understand correctly that the rock-on json is just a way to set up docker containers from the graphical user interface?

Where do I see this mapping between rock-on json and the underlying docker-compose.yaml or Dockerfile?

In order to create scalable containers (and not all-in-ones), the use of network helps creating multi-container environments.
I don't quite understand why one should not follow Docker best-practices when it comes to networking?

Best
Alex

**Comment by FroggyFlox** [comment]
Hi @baforgood,

Thanks for your interest in this issue; I'll try to answer a few of your questions as I've been working on the Rock-On side of Rockstor lately, precisely to improve the setup and configuration of multi-containers  and multi-Rock-Ons environments. I actually just submitted a PR yesterday to allow adding container labels directly from the webUI (#1999), in the idea of easing communication between containers.

I completely agree with our need to make better use of Docker networks and as you can see above, we already have some discussion on how to best implement that. @Luke-Nukem already made a detailed proposition on how we could do that and we started building upon this. Now that I'm on the finishing side of the labels-support feature, I am trying to focus my attention on this. Your timing is thus perfect if you'd like to provide feedback on the discussion above.

Regarding resources and documentation on how Rock-ons function, the [Rockstor documentation](http://rockstor.com/docs/docker-based-rock-ons/overview.html) is always a good starting point, and you will find detailed description of the Rock-On definition file in the r[ockon-registry repo README](https://github.com/rockstor/rockon-registry#what-is-the-structure-of-a-rock-on-profile-file). Please note that we have two additional pending PRs in this repository describing the structure of the recently-added support for `devices` ([here](https://github.com/rockstor/rockon-registry/pull/160)) and `command_arguments` ([here](https://github.com/rockstor/rockon-registry/pull/162)).

Hope this helps,

**Comment by phillxnet** [comment]
@FroggyFlox as our attribution within Web-UI of multiple issues fixed by a single pull request is currently broken, I'm only going to enter a single link in the rpm Changelog (WebUI source). That way folks can at least find this issue via the Changelog within the Web-UI.

Linking to the other associated issues fixed by pr: #2207

Fixes #2003.
Fixes #2013.
Fixes #2009.
